tests/_utils/test_stages.py

```diff
@@ -0,0 +1,224 @@
+
+        pool = multiprocessing.Pool(config.workers or 1)
+
+        jobs = [
+            pool.apply_async(self.run, (path, config, system))
+            for path in config.test_files
+        ]
+        pool.close()
+
+        return [job.get() for job in jobs]
+
+    def _log_proc(
+        self, proc: CompletedProcess[str], test_file: str, verbose: bool
+    ) -> None:
+        msg = f"({self.name}) {test_file}"
+        details = proc.stdout.split("\n") if verbose else None
+        if proc.returncode == 0:
+            LOG(passed(msg, details=details))
+        elif proc.returncode == SKIPPED_RETURNCODE:
+            LOG(skipped(msg))
+        else:
+            LOG(failed(msg, details=details))
+
+
+class CPU(TestStage):
+
+    kind = "cpus"
+
+    def run(
+        self, test_file: str, config: Config, system: System
+    ) -> CompletedProcess[str]:
+        test_path = join(config.root_dir, test_file)
+        stage_args = ["-cunumeric:test"] + next(self.cpu_args(config))
+        file_args = self.file_args(test_file, config)
+
+        cmd = [config.legate_path, test_path]
+        cmd += stage_args + file_args + config.extra_args
+
+        result = system.run(cmd, env=system.env)
+        self._log_proc(result, test_file, config.verbose)
+        return result
+
+    def cpu_args(self, config: Config) -> Iterator[list[str]]:
+        cpus = config.cpus or config.DEFAULT_CPUS
+        yield ["--cpus", str(cpus)]
+
+
+class GPU(TestStage):
+
+    kind = "cuda"
+
+    def run(
+        self, test_file: str, config: Config, system: System
+    ) -> CompletedProcess[str]:
+        test_path = join(config.root_dir, test_file)
+        stage_args = ["-cunumeric:test"] + next(self.gpu_args(config))
+        file_args = self.file_args(test_file, config)
+
+        cmd = [config.legate_path, test_path]
+        cmd += stage_args + file_args + config.extra_args
+
+        result = system.run(cmd, env=system.env)
+        self._log_proc(result, test_file, config.verbose)
+        return result
+
+    def gpu_args(self, config: Config) -> Iterator[list[str]]:
+        gpus = config.gpus or config.DEFAULT_GPUS
+        yield ["--gpus", str(gpus)]
+
+
+class OMP(TestStage):
+
+    kind = "openmp"
+
+    def run(
+        self, test_file: str, config: Config, system: System
+    ) -> CompletedProcess[str]:
+        test_path = join(config.root_dir, test_file)
+        stage_args = ["-cunumeric:test"] + self.omp_args(config)
+        file_args = self.file_args(test_file, config)
+
+        cmd = [config.legate_path, test_path]
+        cmd += stage_args + file_args + config.extra_args
+
+        result = system.run(cmd, env=system.env)
+        self._log_proc(result, test_file, config.verbose)
+        return result
+
+    def omp_args(self, config: Config) -> list[str]:
+        return [
+            "--omps",
+            str(config.omps),
+            "--ompthreads",
+            str(config.ompthreads),
+        ]
```

Comment:
FMI will these ever need to be "sharded" in the same way that will be desired for CPU/GPU? My speculation is that the cmdlin-supplied values will always just be used as-is, but I wanted to check.